src/reconstruct/plm.cpp

```diff
@@ -28,226 +28,234 @@
 
 namespace parthenon {
 
-//----------------------------------------------------------------------------------------
-//! \fn Reconstruction::PiecewiseLinearX1()
-//  \brief
+namespace {
 
-void Reconstruction::PiecewiseLinearX1(const int k, const int j, const int il,
-                                       const int iu, const ParArrayND<Real> &q,
-                                       ParArrayND<Real> &ql, ParArrayND<Real> &qr) {
-  auto pmb = GetBlockPointer();
-  auto &coords = pmb->coords;
-  // set work arrays to shallow copies of scratch arrays
-  ParArrayND<Real> &qc = scr1_ni_, &dql = scr2_ni_, &dqr = scr3_ni_, &dqm = scr4_ni_;
-  const int nu = q.GetDim(4) - 1;
+enum class Dimension { X1, X2, X3 };
+
+struct DataInternal {
+  const int k, j, il, iu, nu;
+  const ParArrayND<Real> &q;
+  ParArrayND<Real> &qc;
+  ParArrayND<Real> &ql;
+  ParArrayND<Real> &qr;
+
+  ParArrayND<Real> &dql;
+  ParArrayND<Real> &dqr;
+  ParArrayND<Real> &dqm;
+  const Coordinates_t &coords;
+  DataInternal(const int k, const int j, const int il, const int iu, const int nu,
+               const ParArrayND<Real> &q, ParArrayND<Real> &qc, ParArrayND<Real> &ql,
+               ParArrayND<Real> &qr, ParArrayND<Real> &dql, ParArrayND<Real> &dqr,
+               ParArrayND<Real> &dqm, const Coordinates_t &coords)
+      : k(k), j(j), il(il), iu(iu), nu(nu), q(q), qc(qc), ql(ql), qr(qr), dql(dql),
+        dqr(dqr), dqm(dqm), coords(coords) {}
+};
+
+void init_dql_and_dqr_and_qc_(DataInternal &data, const Dimension dimension) {
+  int delta_i = 0;
+  int delta_j = 0;
+  int delta_k = 0;
+  if (dimension == Dimension::X1) {
+    delta_i = 1;
+  } else if (dimension == Dimension::X2) {
+    delta_j = 1;
+  } else {
+    delta_k = 1;
+  }
 
   // compute L/R slopes for each variable
-  for (int n = 0; n <= nu; ++n) {
+  const int j = data.j;
+  const int k = data.k;
+  for (int n = 0; n <= data.nu; ++n) {
 #pragma omp simd
-    for (int i = il; i <= iu; ++i) {
+    for (int i = data.il; i <= data.iu; ++i) {
       // renamed dw* -> dq* from plm.cpp
-      dql(n, i) = (q(n, k, j, i) - q(n, k, j, i - 1));
-      dqr(n, i) = (q(n, k, j, i + 1) - q(n, k, j, i));
-      qc(n, i) = q(n, k, j, i);
+      data.dql(n, i) =
+          (data.q(n, k, j, i) - data.q(n, k - delta_k, j - delta_j, i - delta_i));
+      data.dqr(n, i) =
+          (data.q(n, k + delta_k, j + delta_j, i + delta_i) - data.q(n, k, j, i));
+      data.qc(n, i) = data.q(n, k, j, i);
```

Comment:
This is a host-side loop, but `data.q` lives on device. If compiled for GPU, this will die at runtime.